### PR TITLE
Add Hashing and Equality typeclasses.

### PR DESCRIPTION
**Pull request by axel22**
Modify TrieMap to use hashing and equality.

Modify serialization in TrieMap appropriately.

Review by @jsuereth @retronym

### CODE REVIEW

**Comment by axel22** [comment]
Review by @jsuereth

File: src/library/scala/Equality.scala [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+/** `Equality` is a trait whose instances each represent a strategy for determining
+  * equality of instances of a type.
+  *
+  * `Equality`'s companion object defines a default equality for all
+  * objects - it calls their `==` method.
+  *
+  * @since 2.10
+  */
+@annotation.implicitNotFound(msg = "No implicit Equality defined for ${T}.")
+trait Equality[T] extends Serializable {
```
Comment (retronym) [inline src/library/scala/Equality.scala]: We've already got `scala.math.Equiv`.

Comment (axel22) [inline src/library/scala/Equality.scala]: True - we should use that. We should also make that one Serializable.

**Comment by axel22** [comment]
Commit range updated.

Review by @retronym

File: src/library/scala/Equality.scala (hunk 1 above)
Comment (retronym) [inline src/library/scala/Equality.scala]: This is a pretty tricky area: you can't make these guys contravariant, as Paul's thread highlights; it's not clear how to associate compatible Eq/Hash; type class based equality doesn't play nicely with subtyping. Best not to be too hasty with this change, IMO.

Comment (axel22) [inline src/library/scala/Equality.scala]: I agree this needs to be worked out. However, we currently have a request to use custom hashing on TrieMaps. I think we should at least enable the TrieMaps to take custom Hashing objects for now.
Given that we don't do this with the rest of the collections, it should not be hard to switch to a different strategy once we figure it out.

File: src/library/scala/collection/concurrent/TrieMap.scala [hunk 2]
```diff
@@ -807,7 +823,16 @@ extends scala.collection.concurrent.Map[K, V]
     val r = RDCSS_READ_ROOT()
     if (!RDCSS_ROOT(r, r.gcasRead(this), INode.newRootNode[K, V])) clear()
   }
+  
+  @inline private def mangle(hc: Int): Int = {
```
Comment (jsuereth) [inline src/library/scala/collection/concurrent/TrieMap.scala]: Just curious, is it still a good idea to mangle hash codes?  If I can provide my own implementation, that's something i'd have to keep in mind when designing the new hash....

Comment (axel22) [inline src/library/scala/collection/concurrent/TrieMap.scala]: You have a point there, it would be good to get rid of this.

However, the opposing constraint is that when there is no custom Hashing object, some objects produce poor hashcodes for typical object ranges - for example (n: Int).hashCode == n.
The problem with this is that the upper bits of the hash code are usually always 0. In both concurrent and immutable tries this can cause a slowdown due to several collision levels. In parallel hash tables it causes a similar problem with parallel hash table construction.

So, ideally, we want to mangle the bits when a default is provided, and otherwise not mangle them and hope the user knows what he's doing.

File: src/library/scala/math/Hashing.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  */
+@annotation.implicitNotFound(msg = "No implicit Hashing defined for ${T}.")
+trait Hashing[T] extends Serializable {
+  
+  def hashCode(x: T): Int
+  
+}
+
+
+object Hashing {
```
Comment (jsuereth) [inline src/library/scala/math/Hashing.scala]: Note:  `MurmurHash`  exists in `scala.util`. 

(a) Should we provide a default `MurmurHash` instance?  It's an instance of `T => Int`
(b) Should we adapt `MurmurHash` to extends `Hashing`?
(c) Given that we have two classes, should they both be in the same location?  `scala.math.hash`? `scala.util.hash`? `scala.math`? `scala.util`?

We don't have to move anything round in this CL, but we should make a decision on where we want hashing utilities.

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 2 above)
Comment (jsuereth) [inline src/library/scala/collection/concurrent/TrieMap.scala]: What if we figure out if a default is provided?   We can control the default typeclass, so let's give it a real class name and throw an isInstanceOf before mangling?   Either that or we can wrap the typeclass on the way _in_ to the hash map if it's an instance of default....

I agree, keeping the mangling is good for the default.

File: src/library/scala/math/Hashing.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  
+  implicit def default[T] = new Hashing[T] {
```
Comment (jsuereth) [inline src/library/scala/math/Hashing.scala]: Note:  I wish we could use implicit class here...  That would make it simpler to do an isInstanceOf later....

File: src/library/scala/math/Hashing.scala (hunk 3 above)
Comment (axel22) [inline src/library/scala/math/Hashing.scala]: I've put the `Hashing` in `math` because I saw `Ordering` there. But this might not have been the best decision - `scala.util` may be a better option.

Comment (jsuereth) [inline src/library/scala/math/Hashing.scala]: I could go either way, but I think we should pick one.  I'm fine with either `math` or `util`.  I prefer _NOT_ having any packages called `util` because they become an amalgamation of junk.   How bout we keep them either in `scala.math` or move both to `scala.util.hashing` ?

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 2 above)
Comment (axel22) [inline src/library/scala/collection/concurrent/TrieMap.scala]: To avoid `isInstanceOf` on every `computeHash`, I could replace the `hashingobj` with a custom internal `Hashing` which does the mangling if it's a class returned by `Hashing.default`, yes.

File: src/library/scala/math/Hashing.scala (hunk 3 above)
Comment (retronym) [inline src/library/scala/math/Hashing.scala]: Another option: don't integrate `Hash` into the type class hierarchy, but just add the `areEqual` method to it directly.

File: src/library/scala/collection/concurrent/TrieMap.scala (hunk 2 above)
Comment (jsuereth) [inline src/library/scala/collection/concurrent/TrieMap.scala]: +1 ...

File: src/library/scala/math/Hashing.scala (hunk 3 above)
Comment (axel22) [inline src/library/scala/math/Hashing.scala]: I'd put both `Hashing` and `MurmurHash` in `scala.util.hashing`, perhaps, and put deprecated aliases in `scala.util` for `MurmurHash`.

As for the type class hierarchy - currently `Hashing` is not extending `Equiv`. If such a thing is problematic, I do not think we have to do it yet.
The only current use-case for `Hashing` are `TrieMap`s and the `areEqual` method is not necessary there. Do you think we should have it on `Hashing` otherwise?

File: src/library/scala/math/Hashing.scala (hunk 4 above)
Comment (axel22) [inline src/library/scala/math/Hashing.scala]: Yep, would be cool.

File: src/library/scala/math/Hashing.scala (hunk 3 above)
Comment (axel22) [inline src/library/scala/math/Hashing.scala]: I created the `scala.util.hashing` package, but did not yet move `MurmurHash` there. We can do this in a separate commit.
Btw, `MurmurHash` is deprecated in favour of `MurmurHash3` - which does not inherit anything yet. Perhaps it should inherit `Hashing`.

**Comment by axel22** [comment]
Do we close and re-pull here?

File: src/library/scala/math/Hashing.scala (hunk 3 above)
Comment (jsuereth) [inline src/library/scala/math/Hashing.scala]: Yeah, let's do that in a separate commit.  What you have now is exciting and a good first step.

**Comment by jsuereth** [comment]
Did you run a build locally?   I'm fine merging what's there now.  If you spawn Lukas's build bot and link the result, that's good enough for me too.

Otherwise, you would have to close and re-pull to get the bot to run again.

**Comment by axel22** [comment]
## Maybe i should close and repull - i didn't rebuild after the 3rd commit.

Aleksandar Prokopec
LAMP, IC, EPFL

Sent from my Android phone with K-9 Mail. Please excuse my brevity.

Josh Suereth reply@reply.github.com wrote:

Did you run a build locally? I'm fine merging what's there now. If you spawn Lukas's build bot and link the result, that's good enough for me too.

Otherwise, you would have to close and re-pull to get the bot to run again.

---
